### PR TITLE
feat: implement ECharts pie chart

### PR DESCRIPTION
**Pull request by villebro**
🏆 Enhancements
Implement ECharts Pie Chart with same controls as legacy Pie Chart.

Needs to be rebased once `superset-ui/core@0.15.0` lands in `superset-frontend`.

### AFTER

image

### BEFORE

image

### CHART OPTIONS

pie_anim

### CODE REVIEW

**Comment by ktmud** [comment]
Wow, looks so much nicer!

**Comment by villebro** [comment]
@ktmud @rusackas this now works with `0.15` and is ready for re-review

File: plugins/plugin-chart-echarts/src/Pie/controlPanel.ts [hunk 1]
```diff
@@ -17,19 +17,101 @@
...
+            },
+          },
+          {
+            name: 'labels_outside',
+            config: {
+              type: 'CheckboxControl',
+              label: t('Put labels outside'),
+              default: true,
+              renderTrigger: true,
+              description: t('Put the labels outside the pie?'),
```
Comment (ktmud) [inline plugins/plugin-chart-echarts/src/Pie/controlPanel.ts:110]:
```suggestion
              description: t('Put the labels outside of the pie?'),
```

File: plugins/plugin-chart-echarts/src/Pie/controlPanel.ts [hunk 2]
```diff
@@ -17,19 +17,101 @@
...
+            },
+          },
+          {
+            name: 'show_legend',
+            config: {
+              type: 'CheckboxControl',
+              label: t('Legend'),
+              renderTrigger: true,
+              default: true,
+              description: t('Whether to display the legend (toggles)'),
```
Comment (ktmud) [inline plugins/plugin-chart-echarts/src/Pie/controlPanel.ts:85]:
```suggestion
              description: t('Whether to display a legend for the chart'),
```

File: plugins/plugin-chart-echarts/src/Pie/controlPanel.ts [hunk 3]
```diff
@@ -17,19 +17,101 @@
...
+            name: 'number_format',
+            config: {
+              type: 'SelectControl',
+              freeForm: true,
+              label: t('Number format'),
+              renderTrigger: true,
+              default: 'SMART_NUMBER',
+              choices: D3_FORMAT_OPTIONS,
+              description: `${t('D3 format syntax: https://github.com/d3/d3-format')} ${t(
+                'Only applies when the "Label Type" is not set to a percentage.',
```
Comment (ktmud) [inline plugins/plugin-chart-echarts/src/Pie/controlPanel.ts:62]:
```suggestion
                'Only applies when "Label Type" is not set to show percentages.',
```

**Review by ktmud — COMMENTED** [review]
LGTM with some nits

File: plugins/plugin-chart-echarts/src/Pie/index.ts [hunk 4]
```diff
@@ -44,7 +38,11 @@ export default class EchartsPieChartPlugin extends ChartPlugin {
       buildQuery,
       controlPanel,
       loadChart: () => import('./EchartsPie'),
-      metadata,
+      metadata: new ChartMetadata({
+        description: 'Pie chart using ECharts',
+        name: t('EchartsPie'),
```
[the remark this answers is not in the document]
Comment (villebro) [reply plugins/plugin-chart-echarts/src/Pie/index.ts:43]: ..also typo.

File: plugins/plugin-chart-echarts/src/Pie/transformProps.ts [hunk 5]
```diff
@@ -16,60 +16,96 @@
...
+    pieLabelType = 'value',
+    showLabels = true,
+    showLegend = false,
+  } = formData as PieChartFormData;
+  if (metric === undefined) throw new Error('metric must be defined');
+  const { label: metricLabel } = convertMetric(metric);
 
-  const keys = Array.from(new Set(data.map(datum => datum[groupby[0]])));
+  const keys = data.map(datum => extractGroupbyLabel(datum, groupby));
+  const colorFn = CategoricalColorNamespace.getScale(colorScheme as string);
```
[the remark this answers is not in the document]
Comment (villebro) [reply plugins/plugin-chart-echarts/src/Pie/transformProps.ts:51]: I also changed `metric` and `groupby` as required props, as those are enforced in the frontend.

File: plugins/plugin-chart-echarts/src/Pie/transformProps.ts [hunk 6]
```diff
@@ -16,60 +16,96 @@
...
+      itemStyle: {
+        color: colorFn(name),
+      },
     };
   });
 
+  const formatter = (params: { name: string; value: number; percent: number }) =>
+    formatPieLabel({ params, numberFormatter, pieLabelType });
+
   const echartOptions = {
```
[the remark this answers is not in the document]
Comment (villebro) [reply plugins/plugin-chart-echarts/src/Pie/transformProps.ts:68]: Oh nice 👍

File: plugins/plugin-chart-echarts/src/Pie/controlPanel.ts [hunk 7]
```diff
@@ -17,19 +17,101 @@
...
+          {
+            name: 'show_labels',
+            config: {
+              type: 'CheckboxControl',
+              label: t('Show Labels'),
+              renderTrigger: true,
+              default: true,
+              description: t(
+                'Whether to display the labels. Note that the label only displays when the the 5% ' +
+                  'threshold.',
```
[the remark this answers is not in the document]
Comment (villebro) [reply plugins/plugin-chart-echarts/src/Pie/controlPanel.ts:99]: These were copied straight from the legacy Pie chart, along with any potential typos. I will clean these up but leave the old ones unchanged.

File: plugins/plugin-chart-echarts/src/utils/formatters.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { getNumberFormatter, NumberFormats, NumberFormatter } from '@superset-ui/core';
+import { EchartsPieLabelType } from '../Pie/types';
```
[the remark this answers is not in the document]
Comment (villebro) [reply plugins/plugin-chart-echarts/src/utils/formatters.ts:20]: Yeah, I'll just move it into `Pie/transformProps.ts` and break it out later if another chart needs it.

**Comment by villebro** [comment]
@ktmud @rusackas ready for round 2.

**Review by ktmud — APPROVED** [review]

**Review by rusackas — APPROVED** [review]
Thanks for all the follow through on this @villebro and @ktmud. Looking good!

**Pull request merged**